ocaml/xapi/helpers.ml

```diff
@@ -1110,31 +1110,66 @@ let assert_is_valid_ip kind field address =
   if not (is_valid_ip kind address) then
     raise Api_errors.(Server_error (invalid_ip_address_specified, [field]))
 
+module type AbstractIpaddr = sig
+  type t
+
+  module Prefix : sig
+    type addr = t
+
+    type t
+
+    val of_string : string -> (t, [> `Msg of string]) result
+
+    val address : t -> addr
+
+    val bits : t -> int
+  end
+
+  val to_string : t -> string
+end
+
 let parse_cidr kind cidr =
-  try
-    let address, prefixlen = Scanf.sscanf cidr "%s@/%d" (fun a p -> (a, p)) in
-    if not (is_valid_ip kind address) then (
-      error "Invalid address in CIDR (%s)" address ;
-      None
-    ) else if
-        prefixlen < 0
-        || (kind = `ipv4 && prefixlen > 32)
-        || (kind = `ipv6 && prefixlen > 128)
-      then (
-      error "Invalid prefix length in CIDR (%d)" prefixlen ;
-      None
-    ) else
+  let pick_module = function
+    | `ipv4 ->
+        (module Ipaddr.V4 : AbstractIpaddr)
+    | `ipv6 ->
+        (module Ipaddr.V6)
+  in
+  let module AddrParse = (val pick_module kind) in
+  match AddrParse.Prefix.of_string cidr with
+  | Ok ip_t ->
+      let address = AddrParse.Prefix.address ip_t |> AddrParse.to_string in
+      let prefixlen = AddrParse.Prefix.bits ip_t in
       Some (address, prefixlen)
-  with _ ->
-    error "Invalid CIDR format (%s)" cidr ;
-    None
+  | Error e ->
+      let msg = match e with `Msg str -> str | _ -> "" in
+      error "Invalid address in CIDR (%s). %s" cidr msg ;
+      None
 
 let assert_is_valid_cidr kind field cidr =
-  if parse_cidr kind cidr = None then
+  if
+    match kind with
+    | `ipv4or6 ->
+        parse_cidr `ipv4 cidr = None && parse_cidr `ipv6 cidr = None
+    | `ipv4 ->
+        parse_cidr `ipv4 cidr = None
+    | `ipv6 ->
+        parse_cidr `ipv6 cidr = None
+  then
     raise Api_errors.(Server_error (invalid_cidr_address_specified, [field]))
 
 let assert_is_valid_ip_addr kind field address =
-  if (not (is_valid_ip kind address)) && parse_cidr kind address = None then
+  if
+    (not (is_valid_ip kind address))
+    &&
+    match kind with
+    | `ipv4or6 ->
+        parse_cidr `ipv4 address = None && parse_cidr `ipv6 address = None
+    | `ipv4 ->
+        parse_cidr `ipv4 address = None
+    | `ipv6 ->
+        parse_cidr `ipv6 address = None
+  then
```

Comment:
is this the same as the body of `assert_is_valid_cidr`? How about abstracting these two out?

Reply:
Was trying to do this one, @contificate, but misremembered the keywords. Does this `as kind` capture both `ipv4` and `ipv6`?

Reply:
`kind` is not a keyword - it's just the name that is used in the `match` expression. The `as` expression captures the entire pattern - so will bind whatever is matched.